feat: unified data   validation framework for EEBUS use cases

DerAndereAndi opened the pull request:
## Summary

Introduces a generic validation framework and applies it to the EG LPC, MA MGCP, and MA MPC use cases, so data returned by the public API methods is verified against EEBUS spec requirements before it reaches consumers.

The bug-fix portion of the original PR (DeviceConfiguration operator-precedence fix + event-firing consistency across the affected handlers) has been split into **#215** and should land first.

## What's included

### Framework (`usecases/internal/`)

- `validation.go` — generic `Validator[T]` / `BaseValidator[T]` with composable rules (`RequireField`, `RequireScaledNumber`, `ValidateRange`, `ValidateMinMax`, `ValidateEnum`, `ValidateSliceNotEmpty`, `CombineRules`, `ConditionalRule`, `ValidateCustom`).
- `measurement.go` — `MeasurementValidator` with measurement-specific rules (`RequireMeasurementId`, `RequireMeasurementValue`, `RequireValueType`, `RequireValueSource`, `RequireValueSourceMandatory`, `ValidateValueState`), plus `SkipValueState` implementing MGCP-003 ("values with state `error` or `outOfRange` SHALL be ignored").
- `MeasurementPhaseSpecificDataForFilter` now takes a validator and skips individual failing measurements instead of aborting the whole call.
- `VALIDATION.md` — developer guide with usage examples.

### Use-case validators

- **EG LPC** — `LoadControlLimitData`, `DeviceConfigurationKeyValue`, `ElectricalConnectionCharacteristic` validators; wired into `ConsumptionLimit`, `FailsafeConsumptionActivePowerLimit`, `ConsumptionNominalMax`.
- **MA MGCP** — scenario-specific validators for Power, Energy, Current, Voltage, Frequency (incl. MGCP-003 skip behavior).
- **MA MPC** — scenario-specific validators for Power, Energy, Current, Voltage, Frequency.

## Behavior changes

- MPC/MGCP: when no measurement passes validation, both now consistently return `api.ErrDataNotAvailable` (previously divergent between the two use cases).
- MPC: drops the 45–65 Hz frequency-range check — not part of the MPC spec.

## Notes

- No public API changes.
- Depends on **#215**.
- Earlier review feedback from @sthelen-enqs is still applicable to the framework/validator specifics and will be addressed in follow-up commits.

DerAndereAndi committed:
feat: add comprehensive validation system for EEBUS data types

Implement type-safe validation framework using Go generics:

- Add generic validation system in validation.go with BaseValidator
- Support for any SPINE data type with compile-time type safety
- Comprehensive validation rules: RequireField, ValidateRange, ValidateEnum, etc.
- Measurement-specific validation with predefined validators
- Replace MeasurementWithMandatoryData with new validation approach
- Complete GoDoc documentation with usage examples
- Comprehensive test coverage for all validation scenarios
- Markdown documentation guide for developers

Benefits:
- No reflection, direct field access for performance
- Type-safe validators prevent runtime errors
- Composable rules for complex validation scenarios
- Clear error messages with validator context
- Easy extension for new SPINE data types

DerAndereAndi committed:
refactor: implement unified validation framework for EEBUS measurements

- Create generic validation framework with Validator[T] interface
- Implement BaseValidator with composable validation rules
- Add MeasurementValidator for type-safe measurement validation
- Unify validation through MeasurementPhaseSpecificDataForFilter
- Remove non-spec frequency range validation (45-65Hz)
- Add comprehensive test coverage for all validators
- Achieve 100% test coverage for MPC use case

The new validation system ensures spec compliance by:
- Rejecting averageValue types when value is required
- Rejecting empirical sources for energy measurements
- Allowing any ValueState for current measurements
- Validating voltage range (0-1000V) per spec
- Removing frequency range check not in MPC spec

This unified approach reduces code duplication and makes it easier
to maintain consistent validation across all EEBUS use cases.

DerAndereAndi committed:
♻️ refactor: standardize measurement validation behavior and fix ptr function duplication

- Replace local ptr[T any] functions with util.Ptr from spine-go utils package
- Standardize "no valid data found" behavior to return errors consistently for both MPC and MGCP
- Update MGCP test expectations to match consistent error handling
- Remove duplicate utility functions across validator test files
- Ensure both use cases handle data availability scenarios uniformly

DerAndereAndi committed:
✨ feat(eg/lpc): add data validation for EG LPC use case

Add validators for EG LPC (Energy Guard Load Power Consumption) use case to validate
data structures when reading from Controllable Systems:

- LoadControlLimitData: validate required fields (LimitId, Value)
- DeviceConfigurationKeyValueData: validate required fields and value types
- ElectricalConnectionCharacteristicData: validate required fields and characteristic types

Validation follows EEBus LPC specification principles:
- Only validate structural integrity and type requirements
- Trust that CS has validated its own data according to spec
- No arbitrary range validations beyond spec requirements

This ensures data integrity when EG reads from CS devices while following the
unified validation framework pattern established in MGCP and MPC use cases.

DerAndereAndi committed:
🧹 refactor: rename parameters to avoid shadowing built-in min/max functions

- Rename 'min' and 'max' parameters to 'minVal' and 'maxVal' in validation functions
- Fix CodeFactor warnings about redefinition of built-in functions
- No functional changes, improves code maintainability

DerAndereAndi commented:
@sthelen-enqs this should also help for the scenario in your PR https://github.com/enbility/spine-go/pull/52

DerAndereAndi committed:
✅ test: improve test coverage across multiple use case packages

- Increased internal/measurement.go coverage from 69% to 94.1%
- Enhanced eg/lpc package coverage from 94.7% to 95.2%
- Boosted ma/mgcp package coverage from 94.7% to 97.3%
- Achieved 100% coverage for ma/mpc package (up from 97.4%)

Added comprehensive test cases for:
- Measurement validation functions and edge cases
- LPC validator functions and configuration handling
- MGCP/MPC event callbacks for phase-specific measurements
- Error conditions and invalid data scenarios

sthelen-enqs commented on usecases/eg/lpc/public_test.go:
```diff
@@ -417,3 +419,663 @@ func (s *EgLPCSuite) Test_PowerConsumptionNominalMax() {
+	assert.Nil(s.T(), fErr)
+
+	// Test 1: Missing LimitId - should return ErrDataInvalid when validator runs
+	// Note: Data must include Value to pass the `value.Value == nil` check in public.go:59
```
This comment is incorrect, and confusing. GetLimitDataForId will return nil, api.ErrDataNotAvailable because the filter won't match any limits and len(result) will be 0. The check on L59 then short-circuits on err != nil and value.Value will never be checked.
```suggestion
```

sthelen-enqs commented on usecases/eg/lpc/public_test.go:
```diff
@@ -417,3 +419,663 @@ func (s *EgLPCSuite) Test_PowerConsumptionNominalMax() {
+				// LimitId missing - this should trigger validation error
+				IsLimitChangeable: util.Ptr(true),
+				IsLimitActive:     util.Ptr(false),
+				Value:             model.NewScaledNumberType(6000), // Include value to pass initial check
```
```suggestion
				Value:             model.NewScaledNumberType(6000),
```

sthelen-enqs commented on usecases/eg/lpc/public_test.go:
```diff
@@ -417,3 +419,663 @@ func (s *EgLPCSuite) Test_PowerConsumptionNominalMax() {
+	assert.Equal(s.T(), time.Duration(time.Hour*2), data)
+}
+
+func (s *EgLPCSuite) Test_ConsumptionNominalMax_ValidationErrors() {
```
This test case is using the wrong CharacteristicType throughout because the "current" version uses the DeviceType to determine the correct CharacterisicType, but the DeviceType is never initialised in the test suite (and therefore nil).

Could you maybe rebase this PR on fix/use-correct-eg-lpc-lpp-characteristic-types ?

sthelen-enqs reviewed (changes requested):
Hi,

I haven't fully checked through the whole PR, but found a couple of minor nits in the tests that might be nice to fix. I'll check over the rest of the validation later.

general nit: could you run gofmt on the test cases as well?

sthelen-enqs commented on usecases/eg/lpc/public.go:
```diff
@@ -289,6 +303,12 @@ func (e *LPC) ConsumptionNominalMax(entity spineapi.EntityRemoteInterface) (floa
 	}
 
+	// Validate ElectricalConnectionCharacteristicData using EG LPC validator
+	// This ensures nominal consumption values are positive and within reasonable ranges
```
Does it? As far as I can tell, it only validates that a characteristicId is set, and that the characteristicType is reasonable (which it has to be anyway since the filter uses the characterisicType to find the value in the first place.
Realistically this is only checking that CharacteristicId != nil (assuming our filter is functioning correctly).
Neither value nor unit are checked, therefore it's a bit of a stretch to say this ensures values are positive and within reasonable ranges.

sthelen-enqs commented on usecases/eg/lpc/public.go:
```diff
@@ -60,6 +60,13 @@ func (e *LPC) ConsumptionLimit(entity spineapi.EntityRemoteInterface) (
+		resultErr = api.ErrDataInvalid
+		return
+	}
+
```
This does nothing that isn't already done by lines 58-61. It's not technically wrong, it's just dead code that's burning cycles.

sthelen-enqs replied:
Is the intention here to replace the current piecemeal validators with generic validators? In that case this should replace L59-61 and then I'd need to double-check the correctness again.

sthelen-enqs commented on usecases/eg/lpc/public.go:
```diff
@@ -123,6 +130,12 @@ func (e *LPC) FailsafeConsumptionActivePowerLimit(entity spineapi.EntityRemoteIn
+	if err := EGLPCDeviceConfigurationValidator.Validate(data); err != nil {
+		return 0, api.ErrDataInvalid
+	}
+
```
Same as above, except that in addition the comment is very misleading. This code doesn't actually check that power limit values are within reasonable limits. It only checks if KeyId is set (which must be set anyway if we've reached this function cal), and if a value is present that only ScaledNumber and/or Duration are set on it (but not String/Boolean/DateTime).

I'm not sure of the utility of this, and from a validation standpoint it's not even technically correct since:
a) data.Value _must_ be set at this point (the code currently passes when it isn't),
b) I don't think the SPINE/EEBUS spec forbid having additional members of the DeviceConfigurationKeyValueValueType set at this point (such as DateTime, Boolean or String), and
c) if the spec did forbit it, the check would also need to be that  ScaledNumber XOR Duration should be set on it since a failsafeConsumptionActivePowerLimit should not include a duration, and a failsafeDurationMinimum should not contain a scaledNumber

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -28,11 +30,19 @@ func MeasurementPhaseSpecificDataForFilter(
+	if validator == nil {
+		return nil, fmt.Errorf("validator is required")
+	}
+
```
Could we make validator pass-by-value instead of pointer typed? Then it cannot be nil by type guarantee which reduces the risk of errors at runtime and (imo) makes it easier to reason about.

I don't think the struct is large enough that we have to worry about passing it by value, and I think the go compiler should just do the right thing™

Not a blocker in any way, just wondering if we can improve the ergonomics here.

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -28,11 +30,19 @@ func MeasurementPhaseSpecificDataForFilter(
+		if err := validator.Validate(&item); err != nil {
+			// For MGCP-003 compliance, we explicitly handle ErrSkipMeasurement
+			// All other validation errors also result in skipping the measurement
+			continue // Skip invalid measurements, don't fail entire operation
```
Although the reasoning is correct, this error message is confusing as no error is handled explicitly.
It might be nice to have some way of logging/warning/reporting measurements with valueState != normal and probably another one for other validationErrors, but I don't know if we need that in this PR.
```suggestion
			// Due to MGCP-003, and MPC-003 we want to skip measurements with valueState != normal
			// We also want to skip all other validation errors so that we don't fail the entire operation
			// just because the remote contains invalid data
			continue
```

sthelen-enqs commented on usecases/internal/validation.go:
```diff
@@ -0,0 +1,350 @@
+//	if err != nil {
+//		return api.ErrDataNotAvailable
+//	}
+func (v *BaseValidator[T]) ValidateFirst(items []T) (T, error) {
```
I dislike the name of this function. The function is not validating the first element of items[], it's finding the first valid element.
```suggestion
func (v *BaseValidator[T]) FindFirstValidItem(items []T) (T, error) {
```

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+	ptrMeasurements := make([]*model.MeasurementDataType, len(measurements))
+	for i := range measurements {
+		ptrMeasurements[i] = &measurements[i]
+	}
```
I don't see why the conversion here is necessary, but I'm not that bothered by it either.

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+	return func(m *model.MeasurementDataType) error {
+		if expected == "" {
+			return nil // Skip if no expected type specified
+		}
```
This feels like a caller error that we shouldn't be silently ignoring here, if somebody calls this function incorrectly and it isn't doing what they're expecting they should get an error

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+
+// RequireValueSource ensures ValueSource is one of allowed types
+func RequireValueSource(allowed ...model.MeasurementValueSourceType) ValidationRule[*model.MeasurementDataType] {
+	if len(allowed) == 0 {
+		return func(m *model.MeasurementDataType) error { return nil }
+	}
```
This feels like a caller error that we shouldn't be silently ignoring here.

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+			if *m.ValueSource == allowed {
+				return nil
+			}
+		}
```
```suggestion
		// Validate against allowed values
		if slices.Contains(allowed, *m.ValueSource) {
			return nil
		}
```

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+		model.MeasurementValueSourceTypeMeasuredValue,
+		model.MeasurementValueSourceTypeCalculatedValue,
+		model.MeasurementValueSourceTypeEmpiricalValue,
+	)).
```
Afaict valueSource is mandatory for the MU, as this is validator only used in the MA we can (should?) validate that valueSource is set _and_ set correctly.
The same applies for valueSource in the following data points as well.

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+	WithName("Frequency Measurement (Legacy)").
+	WithRule(RequireMeasurementId()).
+	WithRule(RequireMeasurementValue()).
+	WithRule(RequireValueType(model.MeasurementValueTypeTypeValue))
```
I don't see the point in having these here. They're not being used in the stack, if anyone needs them for some reason they can (and should?) add them in their code.
The Voltage one also somehow gained a 0-1000V range which I'm pretty sure wasn't in the old public.go code

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+		model.MeasurementValueSourceTypeEmpiricalValue,
+	)).
+	WithRule(SkipValueState()) // R: Recommended with MGCP-003
+
```
I just noticed that all of these exist in ma/mgcp/validators.go as well. These ones are duplicate and unused?
If so they should be removed.

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+		
+		if len(allowed) == 0 {
+			return nil // Any value is acceptable if no restrictions
+		}
```
I don't like the check for empty allowed here. If you need a field to exist and not be nil you can use RequireField, if you're using RequireValieSourceMandatory my expectation is that you have some restrictions on your valueSource and if you forget them it should probably be an error.

sthelen-enqs commented on usecases/internal/measurement.go:
```diff
@@ -62,16 +72,401 @@ func MeasurementPhaseSpecificDataForFilter(
+		
+		return fmt.Errorf("ValueSource must be one of %v, got %s", allowed, *m.ValueSource)
+	}
+}
```
This function as a whole is confusing. This is just a RequireField().ValidateEnum() chained call that unrolls the chained function call.
I don't mind the unrolling, it's just somewhat arbitrary which functions have an unrolled variant and which don't.
e.g. ValidateRange does not have a ValidateRangeMandatory unrolled variant and instead expects you to call RequireScaledNumber().ValidateRange() manually.

sthelen-enqs commented on usecases/ma/mgcp/validators.go:
```diff
@@ -0,0 +1,114 @@
+		return internal.RequireValueSourceMandatory(mgcpValueSources...)
+	}
+	// For recommended, we validate if present but don't require
+	return internal.RequireValueSource(mgcpValueSources...)
```
Afaict for the GCP valueSource is always Mandatory for the GCP and this should always be internal.RequireValueSourceMandatory(mgcpValueSources...) or if we drop that function to make it more uniform internal.CombineRules(internal.RequireField(...), internal.RequireValueSource(...))

sthelen-enqs commented on usecases/ma/mgcp/validators.go:
```diff
@@ -0,0 +1,114 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(RequireValueSourceMGCP(false)). // Recommended, not mandatory
```
afaict valueSource is always M for the GCP

sthelen-enqs commented on usecases/ma/mgcp/validators.go:
```diff
@@ -0,0 +1,114 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(RequireValueSourceMGCP(false)). // Recommended, not mandatory
```
afaict valueSource is always M for the GCP

sthelen-enqs commented on usecases/ma/mgcp/validators.go:
```diff
@@ -0,0 +1,114 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(RequireValueSourceMGCP(false)). // Recommended, not mandatory
```
afaict valueSource is always M for the GCP

sthelen-enqs commented on usecases/ma/mgcp/validators.go:
```diff
@@ -0,0 +1,114 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(RequireValueSourceMGCP(false)). // Recommended, not mandatory
```
afaict valueSource is always M for the GCP

sthelen-enqs commented on usecases/ma/mgcp/validators.go:
```diff
@@ -0,0 +1,114 @@
+var basicMGCPValidator = internal.NewMeasurementValidator().
+	WithName("MGCP Basic").
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue())
```
```suggestion
```

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	frequencySourceTypes = []model.MeasurementValueSourceType{
+		model.MeasurementValueSourceTypeMeasuredValue,
+		model.MeasurementValueSourceTypeCalculatedValue,
+	}
```
Does it mention anywhere in the spec that empiricalValue isn't valid here?

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(powerSourceTypes...)).
```
Afaict valueSource is always required for the MU and can be made mandatory here

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(energySourceTypes...)).
```
Afaict valueSource is always required for the MU and can be made mandatory here

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(currentSourceTypes...)).
+	WithRule(internal.ValidateValueState("", true)) // Any state required per spec
```
Afaict valueSource is always required for the MU and can be made mandatory here
I, however, see nothing in the MU that makes valueState mandatory for current measurements, I also don't know why there's an exception here allowing "any state".
Because of MPC-003 shouldn't this be an equivalent SkipValueState as in mgcp/validators?

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(powerSourceTypes...)).
+	WithRule(internal.ValidateValueState(model.MeasurementValueStateTypeNormal, false))
```
Because of MPC-003 shouldn't this be an equivalent SkipValueState as in mgcp/validators?

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(energySourceTypes...)).
+	WithRule(internal.ValidateValueState(model.MeasurementValueStateTypeNormal, false))
```
Because of MPC-003 shouldn't this be an equivalent SkipValueState as in mgcp/validators?

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(voltageSourceTypes...)).
+	WithRule(internal.ValidateMeasurementRange(0, 1000)) // 0-1000V per spec
```
I don't see anything in the spec limiting us to 0-1000V and see no reason to do validate this here.

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(frequencySourceTypes...)).
+	WithRule(internal.ValidateValueState(model.MeasurementValueStateTypeNormal, false)) // Reject error states
```
Because of MPC-003 shouldn't this be an equivalent SkipValueState as in mgcp/validators?

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(currentSourceTypes...)).
```
Afaict valueSource is always required for the MU and can be made mandatory here

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithName("MPC Voltage").
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
```
Afaict valueSource is always required for the MU and can be made mandatory here

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(frequencySourceTypes...)).
```
Afaict valueSource is always required for the MU and can be made mandatory here

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+// DEPRECATED: Use MeasurementPhaseSpecificDataForFilter with validators instead
+func getMeasurementValue(measurements []model.MeasurementDataType, validator *internal.MeasurementValidator) (float64, error) {
+	return internal.GetMeasurementValue(measurements, validator)
+}
```
If this is deprecated, can we just get rid of it?

sthelen-enqs commented on usecases/ma/mpc/validators.go:
```diff
@@ -0,0 +1,92 @@
+	WithRule(internal.RequireMeasurementId()).
+	WithRule(internal.RequireMeasurementValue()).
+	WithRule(internal.RequireValueType(model.MeasurementValueTypeTypeValue)).
+	WithRule(internal.RequireValueSource(voltageSourceTypes...)).
```
Because of MPC-003 shouldn't this be an equivalent SkipValueState as in mgcp/validators?

sthelen-enqs reviewed (changes requested):
I have to admit I'm not in love with these changes. The validator logic feels more complicated than necessary, and I fear will make it easier to filter data incorrectly as evidenced by the nits and spec inconsistencies I've found so far.
I, however, do not have a better idea as of right now either.

DerAndereAndi committed:
🐛 fix: ensure events only fire when public methods return valid data

- Fix DeviceConfiguration OR/AND precedence bug in CheckEventPayloadDataForFilter
  (item.KeyId match was OR'd with item.Value != nil instead of AND'd, causing
  the filter to accept any keyvalue item with a value regardless of KeyId)
- Gate 11 event handlers across EG LPC, MA MGCP, and MA MPC on a successful
  public-method call so event notifications stay consistent with the data
  consumers can actually retrieve via the public API
- Update event tests accordingly

DerAndereAndi committed:
fix(eg/lpc): adapt tests to EntityType-based characteristicType

- Test_characteristicType_EdgeCases: mock EntityType() instead of Device()/DeviceType(),
  reflecting the CEM-based check introduced on dev (#191).
- Test_ConsumptionNominalMax_ValidationErrors: use PowerConsumptionNominalMax for the
  monitored EVSE entity instead of ContractualConsumptionNominalMax.
- Test_ConsumptionNominalMax_ErrorCases: correct the missing-CharacteristicId assertion
  from ErrDataNotAvailable to ErrDataInvalid (validator catches it; filter still matches).

Also apply gofmt across the branch.

DerAndereAndi committed:
Address review feedback from @sthelen-enqs

Spec correctness

- MPC: add MPC-003 SkipValueState to all scenario validators (Power,
  Energy, Current, Voltage, Frequency) so measurements with
  valueState=error or outOfRange are ignored.
- MPC & MGCP: make valueSource Mandatory (previously Recommended for
  most scenarios), using the same measuredValue/calculatedValue/
  empiricalValue source list for every scenario.
- MPC: drop the ad-hoc 0-1000V voltage range check (not in spec).

API simplifications

- Pass MeasurementValidator by value, not pointer, so nil is a type
  error rather than a runtime check.
- Rename BaseValidator.ValidateFirst -> FindFirstValidItem: the name
  now describes intent rather than mechanism.
- ValidateEnum and RequireValueSource panic when called with an empty
  allowed list (always a programming error) instead of silently
  no-op'ing; drop the parallel RequireValueSourceMandatory variant.
- Use slices.Contains in RequireValueSource.

Dead code

- Delete the legacy Power/Energy/Current/Voltage/Frequency measurement
  validators in usecases/internal/measurement.go (marked deprecated,
  never referenced outside docs).
- Delete the duplicate MGCP*Validator vars in internal/measurement.go
  that shadowed the canonical ones in usecases/ma/mgcp/validators.go.
- Delete basicMGCPValidator (unused) and the local SkipValueState /
  RequireValueSourceMGCP wrappers in ma/mgcp (use internal directly).
- Delete usecases/eg/lpc/validators.go and validators_test.go: the
  three validators only re-checked fields that the upstream filters
  or existing nil guards already guarantee; the redundant Validate()
  calls in eg/lpc/public.go are removed along with them, and an
  explicit CharacteristicId nil guard is added where relevant.
- Remove VALIDATION.md; it documented the now-deleted legacy
  validators and is out of date.

Tests for the affected packages were reshaped around the new,
smaller API surface.

andig commented on usecases/eg/lpc/public.go:
```diff
@@ -298,16 +285,10 @@ func (e *LPC) ConsumptionNominalMax(entity spineapi.EntityRemoteInterface) (floa
 	if err != nil {
 		return 0, err
-	} else if len(data) == 0 || data[0].Value == nil {
+	} else if len(data) == 0 || data[0].CharacteristicId == nil || data[0].Value == nil {
```
Out of curiosity: shouldn‘t these checks also guard against length != 1?

DerAndereAndi committed:
fix(eg/lpc): guard ConsumptionNominalMax against len != 1

Match the convention used by MPC Power/EnergyConsumed/Frequency and
MGCP Power, which treat a filter result with more than one match the
same as no match: a protocol violation should surface as
ErrDataNotAvailable rather than silently returning the first entry.

Addresses review comment from @andig on public.go:288.
